File: go/gauss.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

const MAXN = 2000 // Tamanho máximo da matriz

// Matrizes e vetores globais
var A [MAXN][MAXN]float64
var B [MAXN]float64
var X [MAXN]float64

func main() {
	// Lendo argumentos da linha de comando
	if len(os.Args) < 2 {
		fmt.Println("Uso: <dimensão da matriz>")
		return
	}

	n, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println("Forneça um número válido para a dimensão da matriz")
		return
	}
	if n < 1 || n > MAXN {
		fmt.Printf("N = %d está fora do limite permitido.\n", n)
		return
	}

	// Inicializando matriz e vetor B
	inicializarMatriz(n)

	// Medindo o tempo de execução
	start := time.Now()
	eliminacaoGaussiana(n)
	duration := time.Since(start)

	// Exibir vetor solução X
	exibirResultado(n)

	fmt.Printf("Tempo de execução: %.3f ms\n", duration.Seconds()*1000)
}

// Função para inicializar a matriz A e vetor B com valores aleatórios
func inicializarMatriz(n int) {
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			A[i][j] = rand.Float64() // Gera um número aleatório
		}
		B[i] = rand.Float64() // Preenchendo vetor B
		X[i] = 0.0
	}
}

// Implementação da eliminação de Gauss
func eliminacaoGaussiana(n int) {
	fmt.Printf("Executando eliminação de Gauss para uma matriz de dimensão %d...\n", n)

	// Etapa de eliminação
	for norm := 0; norm < n-1; norm++ {
		for row := norm + 1; row < n; row++ {
			multiplicador := A[row][norm] / A[norm][norm]
			for col := norm; col < n; col++ {
				A[row][col] -= A[norm][col] * multiplicador
			}
			B[row] -= B[norm] * multiplicador
		}
	}

	// Substituição regressiva para encontrar X
	for row := n - 1; row >= 0; row-- {
		X[row] = B[row]
		for col := row + 1; col < n; col++ {
			X[row] -= A[row][col] * X[col]
		}
		X[row] /= A[row][row]
	}
}

// Exibir vetor solução X (caso seja pequeno o suficiente)
func exibirResultado(n int) {
	if n < 100 {
		fmt.Print("\nX = [")
		for i := 0; i < n; i++ {
			fmt.Printf("%.2f", X[i])
			if i < n-1 {
				fmt.Print("; ")
			} else {
				fmt.Println("]")
			}
		}
	}
}

File: gauss.rs
use std::env;
use rand::Rng;
use std::time::Instant;

const MAXN: usize = 2000; // Tamanho máximo da matriz

// Matrizes e vetores globais
static mut A: [[f64; MAXN]; MAXN] = [[0.0; MAXN]; MAXN];
static mut B: [f64; MAXN] = [0.0; MAXN];
static mut X: [f64; MAXN] = [0.0; MAXN];

fn main() {
    // Lendo argumentos da linha de comando
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Uso: {} <dimensão da matriz>", args[0]);
        return;
    }
    
    let n: usize = args[1].parse().expect("Forneça um número válido para a dimensão da matriz");
    if n < 1 || n > MAXN {
        println!("N = {} está fora do limite permitido.", n);
        return;
    }

    // Inicializando matriz e vetor B
    unsafe {
        inicializar_matriz(n);
    }
    
    // Medindo o tempo de execução
    let inicio = Instant::now();
    unsafe {
        eliminacao_gaussiana(n);
    }
    let duracao = inicio.elapsed();

    // Exibir vetor solução X
    unsafe {
        exibir_resultado(n);
    }
    println!("Tempo de execução: {:.3?} ms", duracao.as_millis());
}

// Função para inicializar a matriz A e vetor B com valores aleatórios
unsafe fn inicializar_matriz(n: usize) {
    let mut rng = rand::thread_rng();
    for i in 0..n {
        for j in 0..n {
            A[i][j] = rng.gen::<f64>(); // Gera um número aleatório
        }
        B[i] = rng.gen::<f64>(); // Preenchendo vetor B
        X[i] = 0.0;
    }
}

// Implementação da eliminação de Gauss
unsafe fn eliminacao_gaussiana(n: usize) {
    println!("Executando eliminação de Gauss...");
    
    // Etapa de eliminação
    for norm in 0..(n - 1) {
        for row in (norm + 1)..n {
            let multiplicador = A[row][norm] / A[norm][norm];
            for col in norm..n {
                A[row][col] -= A[norm][col] * multiplicador;
            }
            B[row] -= B[norm] * multiplicador;
        }
    }
    
    // Substituição regressiva para encontrar X
    for row in (0..n).rev() {
        X[row] = B[row];
        for col in (row + 1)..n {
            X[row] -= A[row][col] * X[col];
        }
        X[row] /= A[row][row];
    }
}

// Exibir vetor solução X (caso seja pequeno o suficiente)
unsafe fn exibir_resultado(n: usize) {
    if n < 100 {
        print!("\nX = [");
        for i in 0..n {
            print!("{:.2}{}", X[i], if i < n - 1 { "; " } else { "]\n" });
        }
    }
}
